keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+This library will implement a simple and familiar API via Custom Resources, and define a _contract_ between app developers and provisioners regarding the Kubernetes resources related to bucket provisioning.
+
+The actual creation of physical buckets and the generation of appropriate credentials belong to each object store provisioner, while the library handles watches on bucket claims, reconciles desired state, and creates and deletes k8s resources.
+
+Since certain object store-specific artifacts, e.g. user name, are needed in order to clean up after a bucket is deleted or revoked, this proposal must allow for some provisioner values to be stored across API calls.
+
+## Design Details
+
+### Custom Resources
+
+(see the [full specs](#custom-resource-details))
+
+The bucket provisioning library utilizes two Custom Resources to abstract an object store bucket and a claim/request for such a bucket.
++ a namespaced _**ObjectBucketClaim**_ (OBC), similar to a PVC, is a formal request for a new or an existing bucket.
++ an _**ObjectBucket**_ (OB), similar to a PV, is the k8s representation of the provisioned bucket and may contain object store-specific data needed to allow the bucket to be de-provisioned. The OB resides in the same namespace as the object store, and is typically not seen by bucket consumers.
+**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.
+
+As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 
+
+OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, bucket lifecycle policies, etc.
+
+### Moving Parts
+
+Before deploying an object bucket consuming app, the OB and OBC CRDs need to be created with RBAC rules granting _get_, _list_, _watch_, and _delete_ verbs.
+**Note:** OBCs do not need delete permission but OBs, secrets, and configmaps do.
+An object store needs to be created and referenced in the storage classes used by OBCs.
+The OBCs needed by the app must exist in the app's namespace.
+
+As is true for dynamic PV provisioning, a bucket provisioner pod needs to be running for each type of object-store existing in the Kubernetes cluster.
+For example, for AWS S3, the developer creates an OBC referencing a storage class which references the S3 store.
+This storage class needs to be created by the admin.
```

Comment:
Note that using ObjectBucketClass instead of StorageClass may allow this class to be stored in a namespaced area along with the OBs